src/components/lib/component/arch/x86_64/cos_asm_simple_stacks.h

```diff
@@ -25,6 +25,57 @@
 
 #define COS_ASM_REQUEST_STACK
 
+#define COS_ULINV_SWITCH_DOMAIN(protdom)			\
+	movl	$protdom,  %ecx;				\
+	addl	%ecx,      %ecx;				\
+	movl	$0b11,     %eax;				\
+	sall	%cl,       %eax;				\
+	notl	%eax;						\
+	andl	$-4,       %eax;				\
+	xor	%rcx,      %rcx;				\
+	xor	%rdx,      %rdx;				\
+	wrpkru;	
+
+/* r13 = tid -> r14 = &stack*/
+#define COS_ULINV_GET_INVSTK					\
+	movabs	$ULK_BASE_ADDR, %r14;				\
+	/* get perthread invstack */				\
+	movq	%r13, %rax;					\
+	shlq    $8, %rax;	    				\
+	/* r14 = &stack */	 				\
+	addq    %rax, %r14; 					
+
+
+/* r14 = &stack */
+#define COS_ULINV_PUSH_INVSTK					\
+	movq    (%r14), %rdx;					\
+	addq    $1,     %rdx;					\
+	shlq    $4,     %rdx;					\
+	/* rdx = &stack_entry */	 			\
+	addq    %r14,   %rdx;					\
+	/* store cap no and sp */				\
+	movq    $0x0123456789abcdef, %rax; 			\
+	movq    %rax, (%rdx); 					\
+	movq    %rsp, 8(%rdx); 					\
+	/* increment tos */					\
```

Comment:
`tos` seems need to have a full name to understand. I might know its meaning, but not sure.